### DISCUSSION TITLE
v7 with Material UI

### DISCUSSION THREAD

**Discussion by shoxter** [react-hook-form/react-hook-form#4731] [Q&A]
When using v6 with Material UI, passing `register` into the `TextField`'s `inputRef` prop was enough to get everything working with RHF. I'm not entirely sure how to migrate over to v7... I've tried calling register and spreading, but it doesn't end up validating as it did before. Additionally, I've tried calling it outside the component and passing in `registerReturnVal.ref` into the `inputRef` property as I did before and still nothing.

What am I missing?

**Comment by bluebill1049** [comment]
Try the following:

```ts
const input = register('test')

<input {...input} inputRef={input.ref} />
```

**Reply by shoxter** [reply]
> const input = register('test')

I'm very grateful for your extremely quick response.

```
const {register, handleSubmit, formState: {errors}} = useForm({
	mode: "all",
	criteriaMode: "all",
	resolver: yupResolver(schema),
	defaultValues: defaultValues ? defaultValues : {}
});
const emailRegister = register('email');

<TextField {...emailRegister} variant="outlined" label="Email" autoComplete="email" autoFocus={true} type="email"
 required={true} margin="normal" inputRef={emailRegister.ref} error={!!errors.email} helperText={errors.email?.message}/>
```

So this is incorrect?

**Reply by shoxter** [reply]
I just saw your edit -- that is in fact what I did. I wonder why the behavior is so erratic. I'll keep looking.

**Reply by bluebill1049** [reply]
looks right to me. 👍

**Reply by bluebill1049** [reply]
For controlled components, I would use Controller/useController instead.

**Reply by shoxter** [reply]
> For controlled components, I would use Controller/useController instead.

Absolutely. I was going to migrate my controlled autocompletes (already using useController with v6 on them) after the rest of my app.

My inputs are currently not validating and randomly clearing out the value with the above code. I may just need a fresh brain to look at it in the AM.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/vigorous-maxwell-45bkj?file=/src/App.js

**Reply by shoxter** [reply]
> https://codesandbox.io/s/vigorous-maxwell-45bkj?file=/src/App.js

Yep! That was it. 

Just a note for anyone who views this in the future -- **you absolutely need to destructure the ref out of what you spread into the `TextField` or you will get wonky behavior.**

Once I destructured the ref out of the `register` calls, all is working as expected. Material UI doesn't like a ref being passed into the `TextField` under the prop name `ref` so it needs to be separated out before you spread it and then passed specifically into the `inputRef` field.

**Reply by bluebill1049** [reply]
because TextInput's ref will actually be the `div` not the input itself. I will make some improvement on the DX.
